### ISSUE TITLE
tutorial relies on deprecated functionnality (`gko::lend`)

### ISSUE THREAD

**Issue by rbourgeois33** [ginkgo-project/ginkgo#1757]
The tutorial: https://github.com/ginkgo-project/ginkgo/wiki/Tutorial-2:-Implement:-Matrices

 Relies on the use of the `gko::lend` function that is deprecated according to my compile output:
 
 `no longer necessary, just pass the object without lend [-Wdeprecated-declarations]
   56 |     generate_stencil_matrix(gko::lend(matrix));
`

As I declared my matrix with:
`auto matrix = gko::share(mtx::create(exec, gko::dim<2>(discretization_points)));`, should I use `generate_stencil_matrix(matrix.get());` ?

Or is there a more up to date tutorial somewhere ?

Thanks in advance,
Rémi

**Comment by gojakuch** [comment]
(pasting this here so that it doesn't get lost being in only in a Slack thread)

# Ginkgo tutorial outline based on my initial questions and confusions

The tutorial outline [here](https://github.com/ginkgo-project/ginkgo/wiki/) is okay except that it promises to build up towards a Poisson solver and never gets close to that, so this idea should either be dropped or extended properly. As for the existing outline:

- Getting Started // I think we can just redirect to INSTALL.md here for the details for the most part, then we'd only need to update it. But it needs to be very detailed, as the user doesn't know anything about Ginkgo at this point. what we can do here is a typical set up and suggestions on how to adjust it for the user's machine with proper cmake variables, so a very simple recipe. we absolutely need to provide a simple ginkgo program that does almost nothing (maybe creates and outputs a matrix or a vector), just so that the user can try to run it and see if ginkgo is installed correctly before continuing with the tutorial

- Implement: Matrices // I'd also love to see some basic operations like copying, comparing, accessing elements, building diagonal matrices, building small matrices with specific values to test smth. In order not to break the tutorial flow that (I assume) is building up a Poisson solver, we can maybe put this part into a separate optional subsection (see some other comments below).

- Implement: Solvers // we should also show how to solve for multiple RHS (but only after showing the 1 RHS variant) and provide a full **copyable** recipe for a very simple example, like a 5x5 real matrix coded directly, so that people can just copy and try. That lowers the entry barrier a lot in my opinion and demonstrates the general structure in the simplest way. Currently, it's a couple of code snippets with text in between which makes it uncopyable.

- Optimize: Measuring Performance // Again, there's no **copyable** example here, this needs to be fixed. But generally, idk if we perhaps need to move this a bit further. not sure if the first thing people want to do is measure the time, maybe let's introduce preconditioners as one of the main features first.

- Optimize: More Suitable Matrix Formats // some additional explanation of the matrix operations could be put here as well, not to overload the first matrix-related page.

- Optimize: Using a Preconditioner // nothing here yet, but again, this should contain a full **copyable** example for maybe 2 different preconditioners for the user to be able to choose.

- Optimize: Using GPUs // nothing here yet, but again, this should contain a full **copyable** example. 

- // also, I'd add some distributed and MPI pages, I think that'd be great. that should also include proper compilation of Ginkgo with MPI (or link to an instruction for that), as I've had some trouble :) Maybe this can also be put as an additional section in the end.

- Bigger picture (additional/optional) // I think our first-time user could benefit from a simple diagram showing inheritance from LinOp and thus similarities between different matrix types and preconditioners, etc. as well as provide a couple of simple methods that can characterise every class so that the user remembers what goes where. really simplified and small, just to put it all into perspective and show a bigger picture. also explain where vectors and matrices are treated differently and where not. I'd definitely love to see that, as this provides clarity to the project structure.

- Solving a Discrete Poisson Equation // Here, once we've already got all the components we can finally build the solver (if we want to). Important part is that this, again, needs to be a fully **copyable** (preferrably) single-file example that the user can copy. Most explanation should be done directly in the code with comments, I think this would reduce the chance of the user skipping through an important note related to a part of the code. This section would check out the following suggestion from DIVIO: "Under your instruction, the student will execute a series of actions to achieve some **end**."

- Customize: Loggers // I'm not sure what these customise steps were supposed to be about

- Customize: Stopping Criterions

- Customize: Matrix Formats 

- Customize: Solvers

- Customize: Preconditioners

General suggestions:

Now, there are also some sections outside of the tutorial in the current outline, which are empty. If our goal is to really build a Poisson solver in the tutorials, we can put all the side things there, in order not to distract the user. However, we then need to provide optional links to the corresponding pages throughout the tutorial for further reading.

As I've already mentioned, I think we need to have a lot of explanations done directly in **copyable** code with comments, as this would reduce the chance of the user skipping through an important note related to a part of the code. In this text, **copyable** means that these things should be standalone examples with all the includes, main function definition — everything, as the user needs to be able to quickly follow the copy-paste-try procedure.

As suggested in DIVIO, tutorials need to work. However, it's hard to track that no change in Ginkgo breaks them, so maybe, as an idea, we should put all of these standalone copyable examples from the tutorial into a separate CI check. This way we'd directly see if a PR breaks the tutorial and if it needs to be updated. Maybe it's already done but I think it'd be easier to integrate that once we have actual compilable program examples for almost every part of the tutorial.
